easy_kclpy/kinesis_processor.py

```diff
@@ -106,40 +75,78 @@ def checkpoint(self, checkpointer, sequence_number=None, sub_sequence_number=Non
+        :param int sub_sequence_number: the sub sequence number associated with this record.
+        :param approximate_arrival_timestamp: the time the record arrived
+        """
+        raise NotImplementedError
 
-                A record processor will be shutdown if it loses its lease.  In this case the KCL will terminate the
-                record processor.  It is not possible to checkpoint once a record processor has lost its lease.
-            * Shutdown Reason TERMINATE:
-                **THE RECORD PROCESSOR MUST CHECKPOINT OR THE KCL WILL BE UNABLE TO PROGRESS**
+    def should_update_sequence(self, sequence_number, sub_sequence_number):
+        """
+        Determines whether a new larger sequence number is available
 
-                A record processor will be shutdown once it reaches the end of a shard.  A shard ending indicates that
-                it has been either split into multiple shards or merged with another shard.  To begin processing the new
-                shard(s) it's required that a final checkpoint occurs.
+        :param int sequence_number: the sequence number from the current record
+        :param int sub_sequence_number: the sub sequence number from the current record
+        :return boolean: true if the largest sequence should be updated, false otherwise
+        """
+        return self._largest_seq == (None, None) or sequence_number > self._largest_seq[0] or \
+            (sequence_number == self._largest_seq[0] and sub_sequence_number > self._largest_seq[1])
 
+    def process_records(self, process_records_input):
+        """
+        Called by a KCLProcess with a list of records to be processed and a checkpointer which accepts sequence numbers
+        from the records to indicate where in the stream to checkpoint.
 
-        :param amazon_kclpy.messages.ShutdownInput shutdown_input: Information related to the shutdown request
+        :param amazon_kclpy.messages.ProcessRecordsInput process_records_input: the records, and metadata about the
+            records.
         """
-        try:
-            if shutdown_input.reason == 'TERMINATE':
-                # Checkpointing with no parameter will checkpoint at the
-                # largest sequence number reached by this processor on this
-                # shard id
-                log.info('Was told to terminate, will attempt to checkpoint.')
-                self.checkpoint(shutdown_input.checkpointer, None)
-            else:  # reason == 'ZOMBIE'
-                log.info('Shutting down due to failover. Will not checkpoint.')
-        except:
-            pass
+
+        for record in process_records_input.records:
+            data = record.binary_data
+            seq = int(record.sequence_number)
+            sub_seq = record.sub_sequence_number
+            key = record.partition_key
+            aat = record.approximate_arrival_timestamp
+
+            try:
+
+                self.process_record(data, key, seq, sub_seq, aat)
```

Comment:
When would we process a record that doesn't update the sequence?

Reply:
This is the logic for it:
```
return self._largest_seq == (None, None) or sequence_number > self._largest_seq[0] or \
            (sequence_number == self._largest_seq[0] and sub_sequence_number > self._largest_seq[1])
```

Reply:
My guess is for cases that for whatever reason, the records are not coming in order, and for some reason we get an old records from before. 

That logic basically always checks for the max sequence that has been seen so far and updates it with that.

Reply:
I was surprised too when I saw that logic.
Because mu impression is events for each shard are guaranteed to come in order.